tile_retriever.py

```diff
@@ -0,0 +1,233 @@
+"""
+Module to retrieve Sentinel-2 tiles.
+
+This approach uses STAC for scene discovery based on a date range and a location.
+The least cloudy scene is retrieved as a whole, and passed to a tiler. The tiler
+will create tiles of 256x256 pixels at different resolutions, track their boundaries,
+and centroids.
+"""
+import backoff
+import click
+import numpy
+import planetary_computer
+import pystac_client
+import stackstac
+from pystac_client.exceptions import APIError
+from rasterio.errors import RasterioIOError
+from rasterio.warp import transform_bounds
+from requests import ConnectionError, HTTPError
+
+BANDS = ["B02", "B03", "B04", "B05", "B06", "B07", "B08", "B8A", "B11", "B12", "SCL"]
+SIZE = 10980
+NODATA = 0
+# SCL classes.
+# 0: NO_DATA
+# 1: SATURATED_OR_DEFECTIVE
+# 2: DARK_AREA_PIXELS
+# 3: CLOUD_SHADOWS
+# 4: VEGETATION
+# 5: NOT_VEGETATED
+# 6: WATER
+# 7: UNCLASSIFIED
+# 8: CLOUD_MEDIUM_PROBABILITY
+# 9: CLOUD_HIGH_PROBABILITY
+# 10: THIN_CIRRUS
+# 11: SNOW
+SCL_FILTER = [0, 1, 3, 8, 9, 10]
+TILE_SIZE = 256
+PIXELS_PER_TILE = TILE_SIZE * TILE_SIZE
+BAD_PIXEL_MAX_PERCENTAGE = 1.0
+RESOLUTION = 10
+RESRES = [20, 40, 80, 160, 300]
+NROFTILES = 40
+TILESFIT = NROFTILES * TILE_SIZE
+COLLECTION = "sentinel-2-l2a"
+STAC_ENDPOINT = "https://planetarycomputer.microsoft.com/api/stac/v1"
+
+
+@backoff.on_exception(
+    backoff.expo,
+    (HTTPError, ConnectionError, RasterioIOError, APIError),
+    max_tries=10,
+    jitter=backoff.full_jitter,
+    raise_on_giveup=True,
+)
+def search(start="2020-03-01", end="2020-06-01", lon=-105.78, lat=35.79, max_items=1):
+    """
+    Search Sentinel-2 STAC items on the MS planetary computer.
+
+    Looks for the least cloudy scene in a time window over a location.
+    """
+    catalog = pystac_client.Client.open(
+        STAC_ENDPOINT,
+        modifier=planetary_computer.sign_inplace,
+    )
+    # Least cloudy scene for the time period.
+    items = catalog.search(
+        intersects=dict(type="Point", coordinates=[lon, lat]),
+        collections=[COLLECTION],
+        datetime=f"{start}/{end}",
+        sortby="eo:cloud_cover",
+        max_items=max_items,
+    ).item_collection()
+
+    return items
+
+
+@backoff.on_exception(
+    backoff.expo,
+    (HTTPError, ConnectionError, RasterioIOError, APIError),
+    max_tries=10,
+    jitter=backoff.full_jitter,
+    raise_on_giveup=True,
+)
+def retrieve(items):
+    """
+    Download the data.
+    """
+    stack = stackstac.stack(
+        items, resolution=RESOLUTION, assets=BANDS, dtype="uint16", fill_value=NODATA
+    )
```

Comment:
Discussion on optimizing reads at different resolutions: Instead of resampling to certain resolution, it will be faster (tens-of-miliseconds instead of ~1min) to read from the overviews (see https://github.com/gjoseph92/stackstac/issues/196#issuecomment-1375683468), but this will result in resolutions that are not round integer numbers.

Any thoughts on sticking to fixed resolutions like 10, 20, 60 versus overview-level defined resolutions? I know we discussed offline about aligning the 512x512 chips to what the Cloud-Optimized GeoTIFF is using internally for faster reads.

Side note: The default resampling with `stackstac` is Nearest Neighbour (see https://stackstac.readthedocs.io/en/v0.5.0/api/main/stackstac.stack.html#stackstac.stack.params.resampling), which is ok for optical images. It might be good to explicitly set the resampling algorithm in the code to be clearer.

Reply:
I think we are going for 10m so no overviews necessary, as we stick to highest resolution available. If we decide to use 20m then yes we should try to optimize for the 20m overview if they exist.